### PR TITLE
verify the path in the autoloader

### PR DESCRIPTION
**Pull request by icewind1991**
limit auto loaded files to only the paths we expect code to be in

### CODE REVIEW

**Comment by icewind1991** [comment]
cc @LukasReschke @Xenopathic @DeepDiver1975 @PVince81

**Comment by Xenopathic** [comment]
:+1:

File: lib/autoloader.php [hunk 1]
```diff
@@ -34,13 +34,29 @@ class Autoloader {
...
 	/**
+	 * Autoloader constructor.
+	 *
+	 * @param string[] $validRoots
+	 */
+	public function __construct(array $validRoots) {
+		$this->validRoots = $validRoots;
+	}
+
+	public function addValidRoot($root) {
```
Comment (LukasReschke) [inline lib/autoloader.php]: PHP Doc please

File: lib/autoloader.php [hunk 2]
```diff
@@ -102,6 +118,15 @@ public function findClass($class) {
 		return $paths;
 	}
 
+	protected function isValidPath($fullPath) {
+		foreach ($this->validRoots as $root) {
+			if (substr($fullPath, 0, strlen($root) + 1) === $root . '/') {
+				return true;
+			}
+		}
+		throw new \Exception('Path nor allowed');
```
Comment (LukasReschke) [inline lib/autoloader.php]: /nor/not/

**Comment by PVince81** [comment]
Will this potentially break any existing apps ? (especially thinking about the ones where class name doesn't match file name)

**Comment by icewind1991** [comment]
> Will this potentially break any existing apps ? (especially thinking about the ones where class name doesn't match file name)

It only checks the path that the autoloader comes up with, not the classname so it shouldn't break anything using classmaps

**Comment by LukasReschke** [comment]
What's the state on this?

**Comment by icewind1991** [comment]
@LukasReschke should be good to merge

**Comment by oparoz** [comment]
There is code in `/tests/lib/testcase.php` that some apps use for tests and now things are broken.


>  [Exception]
  Path not allowed: /tests/lib/testcase.php

**Comment by Xenopathic** [comment]
@oparoz Make sure you are calling the bootstrap correctly, it initializes the autoloader path

**Comment by oparoz** [comment]
Thanks for the tip @Xenopathic! I can see that it adds `tests` to the list of valid roots and that fixed my custom bootstrap, but many apps will break since they either use their own bootstrap or link to `lib/base.php` directly (the official way of doing things):
https://github.com/owncloud/ocdev/blob/master/ocdev/plugins/startapp/templates/8.1/app/phpunit.xml

Couldn't that line be added to `lib/base.php` instead?

**Comment by Xenopathic** [comment]
@oparoz The autoloader should be 'locked down' to its minimum set of directories, for maximum security. The fact that no-one has found a security bug in our tests does not mean that there isn't one (we take a lot of liberties in the tests), so it makes sense only to enable the tests subdirectory when we are actually running tests. For that, you have to make sure that the bootstrap is loaded, or you manually add the prefix.

**Comment by oparoz** [comment]
Ah, got it.

I think it would be wise to post something on the mailing list. I'll open an issue on ocdev.

**Comment by oparoz** [comment]
It has started...
https://travis-ci.org/owncloud/mail/jobs/78364764

**Comment by nickvergessen** [comment]
I vote for revert, not loading all the stuff by default anymore causes major problems for apps.
Also loading base.php should be enough. to get that working again.

Also the expected minimum for a change like this would be a mail to the dev list with a code sample how to get it working again, don't you think?

**Comment by DeepDiver1975** [comment]
Revert!

**Comment by oparoz** [comment]
I agree that the process is broken, but the fix is a one-liner and the patch adds extra security to ownCloud. Maybe fire an email on the dev list asap and keep it in?

**Comment by nickvergessen** [comment]
Well test folderis removed while creating the packages, so...

**Comment by icewind1991** [comment]
Don't revert, only impacts test, simple fix

**Comment by nickvergessen** [comment]
"only"

**Comment by fossxplorer** [comment]
Some of my apps are symlinks and have always worked, but it's broken:
apps3 -> ../oc/apps3/

An unhandled exception has been thrown:
exception 'Exception' with message 'Path not allowed: /home/user1/oc/apps3/activity/appinfo/application.php' in /home/user1/oc7/lib/autoloader.php:132

Stack trace:
#0 /home/user1/oc7/lib/autoloader.php(152): OC\Autoloader->isValidPath('/home/user1...')
#1 [internal function]: OC\Autoloader->load('OCA\\Activity\\Ap...')
#2 /home/user1/oc/apps3/activity/appinfo/app.php(26): spl_autoload_call('OCA\\Activity\\Ap...')
#3 /home/user1/oc7/lib/private/app.php(149): require_once('/home/user1...')
#4 /home/user1/oc7/lib/private/app.php(130): OC_App::requireAppFile('activity')
#5 /home/user1/oc7/lib/private/app.php(109): OC_App::loadApp('activity')
#6 /home/user1/oc7/lib/private/console/application.php(56): OC_App::loadApps()
#7 /home/user1/oc7/console.php(67): OC\Console\Application->loadCommands(Object(Symfony\Component\Console\Output\ConsoleOutput))
#8 /home/user1/oc7/occ(11): require_once('/home/user1...')

**Comment by LukasReschke** [comment]
Yes. Symlinks seem not to be resolved. For the meantime you can use https://github.com/owncloud/core/blob/e9e42fff61a922f11a3b1014d810562537950b6a/config/config.sample.php#L621-L635 to work with multiple apps folders.

**Comment by BernhardPosselt** [comment]
Can we revert this?

Even after adjusting my tests this won't work for cli commandline tests. Why not simply ship without the tests directory? Or very simply don't load test cases when we are not in debug mode ....

**Comment by BernhardPosselt** [comment]
Apart from that this makes no sense: All tests in the tests folder that are autoloadable should be wrapped in classes that extend a PHPUnit testcase. Now even if someone loads a test through an attack, **nothing will happen**. 

Also this is half assed: what about test folders in an app directory?

TL;DR: PR makes no sense, fix your tests to not execute code when included (if there even are any) :-1:

**Comment by Xenopathic** [comment]
@BernhardPosselt I thought your problem was about loading your app code when running tests? Therefore the problem lies in loading your app, not in registering the tests directory in the autoloader. This was done as part of #18839 to harden ownCloud against outdated installed-but-not-enabled apps.

**Comment by BernhardPosselt** [comment]
That's fine, but you are also breaking the test suite.

**Comment by BernhardPosselt** [comment]
Another thing: if you are still going for the appinfo/app.php approach, the autoloader needs to be run before including the app.php, otherwise it depends on what app gets executed first and stuff breaks.

**Comment by LukasReschke** [comment]
> Another thing: if you are still going for the appinfo/app.php approach, the autoloader needs to be run before including the app.php, otherwise it depends on what app gets executed first and stuff breaks.

https://github.com/owncloud/core/pull/19039 should do what you need.
